In [2]:
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
import seaborn as sns
from helper_functions import graph

# Parkinson's Disease Classification EDA

Here we would like to uncover a lot about the data as the documentation for the data is rather blank aside from the study this stems from.

In [3]:
raw_df = pd.read_csv("/Users/marko/Parkinson_Classification/data/raw/pd_speech_features.csv", header=1)
raw_df.head()

id  gender      PPE      DFA     RPDE  numPulses  numPeriodsPulses  \
0   0       1  0.85247  0.71826  0.57227        240               239   
1   0       1  0.76686  0.69481  0.53966        234               233   
2   0       1  0.85083  0.67604  0.58982        232               231   
3   1       0  0.41121  0.79672  0.59257        178               177   
4   1       0  0.32790  0.79782  0.53028        236               235   

   meanPeriodPulses  stdDevPeriodPulses  locPctJitter  ...  \
0          0.008064            0.000087       0.00218  ...   
1          0.008258            0.000073       0.00195  ...   
2          0.008340            0.000060       0.00176  ...   
3          0.010858            0.000183       0.00419  ...   
4          0.008162            0.002669       0.00535  ...   

   tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
0                     1.5620                     2.6445   
1                     1.5589                     3.6107   
2                     1.5643                     2.3308   
3                     3.7805                     3.5664   
4                     6.1727                     5.8416   

   tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
0                     3.8686                     4.2105   
1                    23.5155                    14.1962   
2                     9.4959                    10.7458   
3                     5.2558                    14.0403   
4                     6.0805                     5.7621   

   tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
0                     5.1221                     4.4625   
1                    11.0261                     9.5082   
2                    11.0177                     4.8066   
3                     4.2235                     4.6857   
4                     7.7817                    11.6891   

   tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
0                     2.6202                     3.0004   
1                     6.5245                     6.3431   
2                     2.9199                     3.1495   
3                     4.8460                     6.2650   
4                     8.2103                     5.0559   

   tqwt_kurtosisValue_dec_36  class  
0                    18.9405      1  
1                    45.1780      1  
2                     4.7666      1  
3                     4.0603      1  
4                     6.1164      1  

[5 rows x 755 columns]

755 features! A cursory glance yields that it looks like lots of these features were one-hot encoded or time series were placed in separate columns which is quite interesting. Anyhow, continuing with our analysis:

In [4]:
raw_df.describe()

id      gender         PPE         DFA        RPDE   numPulses  \
count  756.000000  756.000000  756.000000  756.000000  756.000000  756.000000   
mean   125.500000    0.515873    0.746284    0.700414    0.489058  323.972222   
std     72.793721    0.500079    0.169294    0.069718    0.137442   99.219059   
min      0.000000    0.000000    0.041551    0.543500    0.154300    2.000000   
25%     62.750000    0.000000    0.762833    0.647053    0.386537  251.000000   
50%    125.500000    1.000000    0.809655    0.700525    0.484355  317.000000   
75%    188.250000    1.000000    0.834315    0.754985    0.586515  384.250000   
max    251.000000    1.000000    0.907660    0.852640    0.871230  907.000000   

       numPeriodsPulses  meanPeriodPulses  stdDevPeriodPulses  locPctJitter  \
count        756.000000        756.000000          756.000000    756.000000   
mean         322.678571          0.006360            0.000383      0.002324   
std           99.402499          0.001826            0.000728      0.002628   
min            1.000000          0.002107            0.000011      0.000210   
25%          250.000000          0.005003            0.000049      0.000970   
50%          316.000000          0.006048            0.000077      0.001495   
75%          383.250000          0.007528            0.000171      0.002520   
max          905.000000          0.012966            0.003483      0.027750   

       ...  tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
count  ...                 756.000000                 756.000000   
mean   ...                  26.237251                  22.840337   
std    ...                  42.220693                  32.626464   
min    ...                   1.509800                   1.531700   
25%    ...                   2.408675                   3.452800   
50%    ...                   5.586300                   7.062750   
75%    ...                  28.958075                  29.830850   
max    ...                 239.788800                 203.311300   

       tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
count                 756.000000                 756.000000   
mean                   18.587888                  13.872018   
std                    25.537464                  20.046029   
min                     1.582900                   1.747200   
25%                     3.354825                   3.077450   
50%                     6.077400                   4.770850   
75%                    21.944050                  13.188000   
max                   121.542900                 102.207000   

       tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
count                 756.000000                 756.000000   
mean                   12.218953                  12.375335   
std                    17.783642                  16.341665   
min                     1.789500                   1.628700   
25%                     2.937025                   3.114375   
50%                     4.300450                   4.741450   
75%                    10.876150                  12.201325   
max                    85.571700                  73.532200   

       tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
count                 756.000000                 756.000000   
mean                   14.799230                  14.751559   
std                    15.722502                  14.432979   
min                     1.861700                   1.955900   
25%                     3.665925                   3.741275   
50%                     6.725700                   7.334250   
75%                    21.922050                  22.495175   
max                    62.007300                  57.544300   

       tqwt_kurtosisValue_dec_36       class  
count                 756.000000  756.000000  
mean                   31.481110    0.746032  
std                    34.230991    0.435568  
min                     2.364000    0.000000  
25%                  

Alright, nothing particularly meaningful or interesting coming from these routine checks. Although I will be interested in looking at say the class of variables indexed by time to see if there's any discrepency. The other thing to note is just simply looking at the features most correlated to the outcome as opposed to the features cited in the paper. The baseline features that were most popular according to the paper below are: Jitter, shimmer, fundamental frequency parameters, harmonicity parameters, RPDE, DFA, and PPE. 

In [6]:
corr_mat = raw_df.drop("class", axis=1).apply(lambda col: col.corr(raw_df["class"],
                                                                   method="spearman"))

# 50 largest correlations in absolute value
largest_corr = corr_mat.iloc[corr_mat.abs().argsort()][-50:].sort_values(ascending=True)
graph.barplot(largest_corr,
              savefig=True, 
              title=None,
              xlab='Spearman Correlation with Class')
plt.savefig("./figs/spearman_corr_to_class.png", bbox_inches="tight")

I see some of the features mentioned in the paper - jitter, fundamental frequency parameters, and quite a lot of the time-series elements are all negatively correlated. However we need to test significance of time trends in the entropy, TKEO, and stdValue features to see if we can perhaps combine these to condense the data down.

In [21]:
def na_pct(col):
    return len(col[col.isna()]) / len(col)

na_pcts = raw_df.apply(na_pct, axis=0)
na_pcts[na_pcts != 0]

Series([], dtype: float64)

There is no missing data in any of the columns surprisingly. We can also check for duplicate IDs. 

In [23]:
raw_df["id"].value_counts()

0      3
173    3
160    3
161    3
162    3
      ..
88     3
89     3
90     3
91     3
251    3
Name: id, Length: 252, dtype: int64

Okay, so it appears as though each patient was scanned three times. Let us check to see if these scans were consistent for all patients - it would be surprising if the outcome for some patients changed over different measurements. Let us see:

In [31]:
# look at variance of outcome - if not zero, then there's another 
id_var = raw_df.groupby("id")["class"].var()
id_var[id_var != 0]

Series([], Name: class, dtype: float64)

That's a good sanity check - there's no variance in the outcomes for data with the same IDs. However, the variances for other features such as "PPE" are nonzero, so these columns are not immediately redundant. 

# Sources

The original paper is given here: https://www.sciencedirect.com/science/article/pii/S1568494618305799?via%3Dihub